File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode1695MaximumErasureValue.cpp
#include "LeetCode1695MaximumErasureValue.h"

#include <algorithm>
#include <unordered_map>
using namespace std;

int LeetCode1695MaximumErasureValue::maximumUniqueSubarray(vector<int>& nums)
{
    int sum{};
    int maxSum{};
    unordered_map<int, int> map{};
    int firstPointer{};
    int secondPointer{};
    int length = nums.size();
    while (secondPointer < length)
    {
        if (!map.contains(nums[secondPointer]))
        {
            map[nums[secondPointer]]= secondPointer;
            sum+= nums[secondPointer];
            maxSum = std::max(sum, maxSum);
            secondPointer++;
            continue;
        }
        for (int i=firstPointer,e=map[nums[secondPointer]];i<=e;i++)
        {
            sum-= nums[i];
            map.erase(nums[i]);
            firstPointer++;
        }
        map[nums[secondPointer]]= secondPointer;
        sum+= nums[secondPointer];
        maxSum = std::max(sum, maxSum);
        secondPointer++;
    }
    return  maxSum;
}

File: LeetCodeProblemsGo/LeetCode1695MaximumErasureValue/LeetCode1695MaximumErasureValue.go
package LeetCode1695MaximumErasureValue

//prefix sum solution O(n) time and O(n) space
func maximumUniqueSubarray(nums []int) int {
	sum := 0
	maxSum := 0
	firstPointer := 0
	secondPointer := 0
	length := len(nums)
	prefixSum := make([]int, length)
	numsMap := make(map[int]int)
	for secondPointer < length {
		if _, exists := numsMap[nums[secondPointer]]; exists {
			if numsMap[nums[secondPointer]] >= firstPointer {
				firstPointer = numsMap[nums[secondPointer]] + 1
			}
		}
		if secondPointer == 0 {
			prefixSum[secondPointer] = nums[secondPointer]
		} else {
			prefixSum[secondPointer] = prefixSum[secondPointer-1] + nums[secondPointer]
		}
		numsMap[nums[secondPointer]] = secondPointer
		sum = prefixSum[secondPointer] - prefixSum[firstPointer] + nums[firstPointer]
		maxSum = max(maxSum, sum)
		secondPointer++
	}
	return maxSum
}

// without prefix sum still O(n) and uses less memory simpler code too
// func maximumUniqueSubarray(nums []int) int {
//     sum := 0
// 	maxSum := 0
// 	firstPointer := 0
// 	secondPointer := 0
// 	length := len(nums)
// 	numsMap := make(map[int]int)
// 	for secondPointer < length {
// 		if _, exists := numsMap[nums[secondPointer]]; exists {
// 			for i := firstPointer; i <= numsMap[nums[secondPointer]]; i++ {
// 				sum -= nums[i]
// 				delete(numsMap, nums[i])
// 				firstPointer++
// 			}
// 		}

// 		numsMap[nums[secondPointer]] = secondPointer
// 		sum += nums[secondPointer]
// 			maxSum = max(maxSum, sum)
// 		secondPointer++
// 	}
// 	return maxSum
// }
